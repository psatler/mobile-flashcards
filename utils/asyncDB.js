import { AsyncStorage } from 'react-native'
import { data } from './mockData'
const FLASHCARDS_DECKS_KEY = 'Flashcards:decks'

const setInitialData = async () => {
    let obj = data;

    await AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(obj))
}


export const getDecks = async () => {
    try {
        let value = await AsyncStorage.getItem(FLASHCARDS_DECKS_KEY);
        if( value !== null){ //we have data
            // console.log('decks', value);
            let parsedObj = JSON.parse(value); //parsing back to object
            return parsedObj;
        } else { //if not, we set initial state
            console.log('setting initial state');
            setInitialData(); //defined above
        }
    } catch (err) {
        console.log('There was an error retrieving decks from local DB: ', err);
    }
}

export const saveDeckTitle = (keyTitle, deckObj) => {
    try {
        return AsyncStorage.mergeItem(FLASHCARDS_DECKS_KEY, JSON.stringify({
            [keyTitle]: deckObj
        }))
        
    } catch (err) {
        console.log(err);
    }
}

export const addCardToDeck = (deckTitle, card) => {
    const key = deckTitle;

    return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
        .then( (results) => {
            const allDecks = JSON.parse(results);
            const deck = allDecks[key];
            // console.log('addCardToDeck- BEFORE', deck.questions)
            deck.questions.push(card)
            // console.log('addCardToDeck', deck.questions)
            // console.log('allDecks', allDecks)

            AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(allDecks)); //updating 

        })
        // const key = deckTitle.split(' ').join('');

}

// 

// const checkIfDataExists = (results) => {
//     return results === null
//       ? setInitialData()
//       : JSON.parse(results)
//   }

// export const retrieveDataFromAsyncStorage = () => {

//     return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
//         .then(checkIfDataExists)
// }



// _setValue = async () => {
//     try {
//       await AsyncStorage.setItem(STORAGE_KEY, 'true');
//     } catch (error) { // log the error
//     }
//   };

//   _loadInitialState = async () => {
//     try {
//       let value = await AsyncStorage.getItem(STORAGE_KEY);
//       if (value === 'true'){
//         this.setState({loaded: 'true'});
//       } else {
//         this.setState({loaded: 'false'});
//         this._setValue();
//       }
//     } catch (error) {
//       this.setState({loaded: 'false'});
//       this._setValue();
//     }
//   };