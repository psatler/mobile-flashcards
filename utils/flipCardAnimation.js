import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Animated
} from 'react-native';

export default class AnimatedBasic extends Component {

    state = {
        animatedValue: new Animated.Value(0),
        listenerValue: 0,
    }

    componentDidMount(){
        const { animatedValue, listenerValue } = this.state;

        //creating a listener to track values changes
        animatedValue.addListener( ( { value } ) => {
            this.setState({
                listenerValue: value,
            })
        })
    }

    flipCard = () => {
        const { animatedValue, listenerValue } = this.state;

        if(listenerValue >= 90){
            Animated.spring(animatedValue,{
                toValue: 0,
                friction: 8,
                tension: 10
              }).start();
        } 
        else {
            Animated.spring(animatedValue,{
                toValue: 180,
                friction: 8,
                tension: 10
              }).start();
        }
    }

    render() {
        const { animatedValue, listenerValue } = this.state;

        const frontAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
                outputRange: ['0deg', '180deg'],
              })
            }
          ],
          //adding opacity so the back face won't be mirrored on Android devices
        //https://github.com/facebook/react-native/issues/1973#issuecomment-262059217
          opacity: animatedValue.interpolate({ //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90],
                outputRange: [1, 0]
            }), 
        }
        const backAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
              //   outputRange: ['180deg', '360deg'],
                outputRange: ['180deg', '0deg']
              }) 
            }
          ],
          opacity: animatedValue.interpolate({  //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90], 
                outputRange: [0, 1] 
            }), 
        }
        
        return (
          <View style={styles.container}>
            <View>
              <Animated.View style={[styles.flipCard, frontAnimatedStyle]}>
                <Text style={styles.flipText}>
                  This text is flipping on the front.
                </Text>
              </Animated.View>
              <Animated.View style={[backAnimatedStyle, styles.flipCard, styles.flipCardBack]}>
                <Text style={styles.flipText}>
                  This text is flipping on the back.
                </Text>
              </Animated.View>
            </View>
            <TouchableOpacity onPress={() => this.flipCard()}>
              <Text>Flip!</Text>
            </TouchableOpacity>
          </View>
        );
      }
    }


const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  flipCard: {
    width: 200,
    height: 200,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'blue',
    backfaceVisibility: 'hidden', //it means when the card rotates out of the way, it won't be visible. When it rotates back, it'll be visible again. Not working on Android though, so it was added an workaround for this above
  },
  flipCardBack: {
    backgroundColor: "red",
    position: "absolute",
    top: 0,
  },
  flipText: {
    width: 90,
    fontSize: 20,
    color: 'white',
    fontWeight: 'bold',
  }
});