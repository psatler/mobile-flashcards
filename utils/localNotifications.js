// import React from 'react'
import { AsyncStorage } from 'react-native'
import { Notifications, Permissions } from 'expo'

const FLASHCARDS_NOTIFICATIONS_KEY = 'Flashcards:localNotifications'


const createNotification = () => {
    return {
        title: 'Do a quiz today!',
        body: 'ðŸ‘‹ Do not forget to study your decks!',
        ios: {
            sound: true,
        },
        android: {
            sound: true,
            priority: 'high',
            sticky: false,
            vibrate: true,
        }
    }
}

export const setLocalNotification = () => {
    //checking if our local notification has already been set
    AsyncStorage.getItem(FLASHCARDS_NOTIFICATIONS_KEY) 
        .then(JSON.parse) //parsing the returned value
        .then( (data) => {
            if(data === null){ //if we haven't any notification scheduled
                Permissions.askAsync(Permissions.NOTIFICATIONS)
                    .then( ({ status }) => {
                        if( status === 'granted' ){
                            Notifications.cancelAllScheduledNotificationsAsync();

                            let tomorrow = new Date();
                            tomorrow.setDate(tomorrow.getDate() + 1) //tomorrow 
                            tomorrow.setHours(10); //scheduling for 10 AM
                            tomorrow.setMinutes(0);

                            Notifications.scheduleLocalNotificationAsync(
                                createNotification(),
                                {
                                    time: tomorrow,
                                    repeat: 'day'
                                }
                            )

                            AsyncStorage.setItem(FLASHCARDS_NOTIFICATIONS_KEY, JSON.stringify(true))
                        }
                    })

            }
        })

}

export const clearLocalNotification = () => {
    return AsyncStorage.removeItem(FLASHCARDS_NOTIFICATIONS_KEY)
      .then(Notifications.cancelAllScheduledNotificationsAsync)
}