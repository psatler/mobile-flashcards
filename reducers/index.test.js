import { data } from '../utils/mockData'
import { 
    RETRIEVE_DECKS,
    ADD_DECK,
    ADD_CARD,
    REMOVE_DECK,
    REMOVE_ALL_DECKS,
} from '../actions'

import { deckReducer } from './'

const newDeck = {"newDeck": {"questions": [], "title": "new Deck"}}
const newCardObj = {
    question: "This is a test question",
    answer: "This is a test answer",
}

describe('Deck Reducer', () => {
    it('should handle initial state', () => {

        expect(deckReducer(undefined, { })) //reducer receiving nothing, with an empty action
            .toEqual({ }); //to equal empty
    });

    it('should handle REMOVE_ALL_DECKS action', () => {
        expect(deckReducer({}, {
            type: REMOVE_ALL_DECKS,
        })).toEqual( {} );
    });

    it('should handle RETRIEVE_DECKS action', () => {

        expect(deckReducer( {}, {
            type: RETRIEVE_DECKS,
            payload: data,
        })).toEqual( {
            ...data,
        });
    });

    it('should handle ADD_DECK action', () => {
        expect(deckReducer( {}, {
            type: ADD_DECK,
            payload: newDeck,
        })).toEqual({
            ...newDeck,
        })
    });

    it('should handle ADD_CARD action', () => {

        expect(deckReducer( { ...newDeck }, {
            type: ADD_CARD,
            payload: newCardObj,
            title: "newDeck" , //inserting in the newDeck const above to match snapshot
        }))
        .toMatchSnapshot(); //so it iniatially has no cards. Afterwards, it'll have one (newCardObj)
    });

    it('should handle REMOVE_DECK action', () => {

        expect(deckReducer( { ...data }, {
            type: REMOVE_DECK,
            payload: "React", 
        }))
        .toMatchSnapshot(); //it should remove the "React" deck from the mockData
    });
})