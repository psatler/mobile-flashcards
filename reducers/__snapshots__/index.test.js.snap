// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck Reducer should handle ADD_CARD action 1`] = `
Object {
  "newDeck": Object {
    "createdAt": undefined,
    "image": undefined,
    "questions": Array [
      Object {
        "answer": "This is a test answer",
        "question": "This is a test question",
      },
    ],
    "title": "new Deck",
  },
}
`;

exports[`Deck Reducer should handle REMOVE_DECK action 1`] = `
Object {
  "JavaScript": Object {
    "questions": Array [
      Object {
        "answer": "The combination of a function and the lexical environment within which that function was declared.",
        "question": "What is a closure?",
      },
    ],
    "title": "JavaScript",
  },
}
`;
